### DISCUSSION TITLE
Sentry Support has Landed 🤯

### DISCUSSION THREAD

**Discussion by kentcdodds** [epicweb-dev/epic-stack#154] [Announcements]
Huge thank you to @scefali for https://github.com/epicweb-dev/epic-stack/pull/40!

Please checkout the docs to learn how to get this setup: https://github.com/epicweb-dev/epic-stack/blob/main/docs/monitoring.md

Read the decision document here: https://github.com/epicweb-dev/epic-stack/blob/main/docs/decisions/015-monitoring.md

**Comment by cayblood** [comment]
The decision document doesn't include anything about how to generate sourcemaps during deploy so that Sentry can show proper context for errors, and I don't see anything in the github actions that seems to do that. Is that also configured properly, or do we still need to do that on our own? Thx.

**Reply by kentcdodds** [reply]
That's set here: https://github.com/epicweb-dev/epic-stack/blob/512f920bbc5da0dbcfed8e4bc26a3eb174fef60c/package.json#L15

**Reply by cayblood** [reply]
Thanks! Where do the sourcemaps get uploaded to Sentry?

**Reply by kiliman** [reply]
Here's my `Dockerfile`, based on _Indie Stack_, not _Epic Stack_
```bash
# base node image
FROM node:16-bullseye-slim as base

# set for base and all layer that inherit from it
ARG COMMIT_SHA
ARG APP_VERSION
ARG SERVER_URL
ARG SENTRY_AUTH_TOKEN
ARG SENTRY_ORG
ARG SENTRY_PROJECT

ENV COMMIT_SHA=$COMMIT_SHA
ENV SENTRY_AUTH_TOKEN=$SENTRY_AUTH_TOKEN
ENV SENTRY_ORG=$SENTRY_ORG
ENV SENTRY_PROJECT=$SENTRY_PROJECT
ENV VERSION=$APP_VERSION-$COMMIT_SHA

# Install openssl for Prisma
RUN apt-get update && apt-get install -y openssl sqlite3 ca-certificates

WORKDIR /myapp

ADD package.json package-lock.json ./
ADD . .
RUN npm install

ADD prisma .
RUN npx prisma generate

ENV NODE_ENV production
RUN npm run build

# start creating the release
RUN npx sentry-cli releases new "$VERSION"
# upload server sourcemaps
RUN npx sentry-cli releases files "$VERSION" upload-sourcemaps ./ --ignore node_modules --ignore public --url-prefix="/myapp"
# upload client sourcemaps (fully separate them by setting dist to client)
RUN npx sentry-cli releases files "$VERSION" upload-sourcemaps ./public --dist "client"
RUN npx sentry-cli releases finalize "$VERSION"

# remove sourcemaps
RUN rm -rf build/*.map public/build/*.map

ENV DATABASE_URL=file:/data/data.db
ENV PORT="8080"
ENV NODE_ENV="production"

# add shortcut for connecting to database CLI
RUN echo "#!/bin/sh\nset -x\nsqlite3 \$DATABASE_URL" > /usr/local/bin/database-cli && chmod +x /usr/local/bin/database-cli

CMD ["npm", "start"]
```

**Reply by kiliman** [reply]
Needless to say, it took some trial and error to get it working. 😔

<img width="648" alt="image" src="https://github.com/epicweb-dev/epic-stack/assets/47168/86c94aa9-61aa-42e5-933f-d14eadcea4e3">

**Reply by kiliman** [reply]
Make sure you initialize Sentry with the same 'version' you use when uploading the sourcemaps. For my app, I set it to `VERSION=APP_VERSION-COMMIT_HASH` where `APP_VERSION=1.0`

```ts
// root.tsx
console.log(
  `🐛 init sentry client: ${ENV.RELEASE_STAGE} v${ENV.VERSION}`,
);
Sentry.init({
  environment: ENV.RELEASE_STAGE, // staging or production
  release: ENV.VERSION,           // same as VERSION for sourcemaps
  dsn: ENV.SENTRY_DSN,
  tracesSampleRate: 1,
  integrations: [
    new Sentry.BrowserTracing({
      routingInstrumentation: Sentry.remixRouterInstrumentation(
        useEffect,
        useLocation,
        useMatches,
      ),
    }),
  ],
});
```
```ts
// entry.server.tsx
function myTracesSampler(samplingContext: SamplingContext) {
  // Don't trace healthcheck or HEAD requests
  const { transactionContext } = samplingContext;
  if (
    transactionContext.name === 'routes/healthcheck/_index' ||
    transactionContext.tags?.method === 'HEAD'
  ) {
    return false;
  } else {
    return 1.0;
  }
}
if (env.SENTRY_NOTIFY) {
  Sentry.init({
    environment: env.RELEASE_STAGE,
    release: `${env.APP_VERSION}-${env.COMMIT_SHA}`,
    dsn: env.SENTRY_DSN,
    tracesSampleRate: 1,
    tracesSampler: myTracesSampler,

    integrations: [new Sentry.Integrations.Prisma({ client: prisma })],
  });
}
```
>NOTE: I had to filter out the `/healthcheck` endpoint since Fly was calling it every 10 seconds and it was blowing up the transaction count.

Here's an example error showing the stack trace on the server
<img width="1076" alt="image" src="https://github.com/epicweb-dev/epic-stack/assets/47168/3148c76a-b68d-49b5-b791-81ab866aa96b">

And here's what it looks like after you upload the sourcemaps
<img width="1217" alt="image" src="https://github.com/epicweb-dev/epic-stack/assets/47168/cd86e87f-dca1-4ecc-9b57-2a10a23e63b1">

<img width="1200" alt="image" src="https://github.com/epicweb-dev/epic-stack/assets/47168/e546c92f-d353-4ecc-a573-a1eb2c3167a5">

**Reply by kentcdodds** [reply]
This is the bit you may be looking for in the Epic Stack: https://github.com/epicweb-dev/epic-stack/blob/main/other/sentry-create-release.js

**Reply by cayblood** [reply]
@kiliman @kentcdodds thanks so much for these responses. That was enough for me to get it working.
